### PR TITLE
Add base support for ATEN Bool tensor type.

### PR DESCRIPTION
**Pull request by dlibenzi**
Added debug build support for TF and C++ tests.
Changed s_copy_() tests as currently remote TF XLA_CPU device has issues.
Fixed TriangularSolve to clear the layouts when adding shape's dimensions.
Fixed device data caching to be per-device.

### CODE REVIEW

File: torch_xla/csrc/tensor.cpp [hunk 1]
```diff
@@ -136,35 +136,59 @@ std::vector<xla::util::Cleanup> LockDevices(const std::set<Device>& devices) {
   return unlocker;
 }
 
-struct TensorHasher {
```
Comment (asuhan) [inline torch_xla/csrc/tensor.cpp]: What are the changes to this file for? They don't look related to the bool tensor.

Comment (dlibenzi) [reply torch_xla/csrc/tensor.cpp]: They started as bool tensors, and my "Let me try that on TPU".
Then hell broke lose and I accumulated them here.

Comment (asuhan) [reply torch_xla/csrc/tensor.cpp]: I'd rather split into two commits if it's not too hard.

Comment (dlibenzi) [reply torch_xla/csrc/tensor.cpp]: Please ...

Comment (asuhan) [reply torch_xla/csrc/tensor.cpp]: I don't think this is an unreasonable ask, but ok.

File: torch_xla/csrc/tensor.cpp [hunk 2]
```diff
@@ -136,35 +136,59 @@ std::vector<xla::util::Cleanup> LockDevices(const std::set<Device>& devices) {
...
+                       TensorComparer>;
+
+  explicit XlaDataCacheArena(size_t max_cache_size)
+      : max_cache_size_(max_cache_size) {}
+
+  XlaDataCache* Get(const Device& device) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    auto it = device_caches_.find(device);
+    if (it == device_caches_.end()) {
+      it = device_caches_.emplace(device, new XlaDataCache(max_cache_size_))
```
Comment (asuhan) [inline torch_xla/csrc/tensor.cpp]: You need to emplace an unique pointer here, this is why the build fails.

Comment (dlibenzi) [reply torch_xla/csrc/tensor.cpp]: Done.
Weird, was working for me on my workstation.

Comment (asuhan) [reply torch_xla/csrc/tensor.cpp]: CircleCI uses a different compiler.
